# Assignment — Week 06 · Day 30 (AM+PM)
**PG Diploma · AI-ML & Agentic AI Engineering · IIT Gandhinagar**  
**Topic:** Logistic Regression — End-to-End ML Pipeline  
**Dataset:** SUV Purchase Dataset

---

## Part A — Concept Application

### Step 1 — Data Loading & Exploration

In [5]:
# Import libraries
import pandas as pd
import numpy as np

# Load the dataset
df = pd.read_csv("C:\\Users\\avish\\Downloads\\suv_data.csv")

# Display first 5 rows
print("First 5 rows:")
df.head()

First 5 rows:


,User ID,Gender,Age,EstimatedSalary,Purchased
0,15624510.0,Male,19.0,19000.0,0.0
1,15810944.0,Male,35.0,20000.0,0.0
2,15668575.0,Female,26.0,43000.0,0.0
3,15603246.0,Female,27.0,57000.0,0.0
4,15804002.0,Male,19.0,76000.0,0.0


In [6]:
# Shape and column names
print("Shape:", df.shape)
print("\nColumn Names:", df.columns.tolist())

Shape: (400, 5)

Column Names: ['User ID', 'Gender', 'Age', 'EstimatedSalary', 'Purchased']


In [7]:
# Data types
print("Data Types:")
print(df.dtypes)

Data Types:
User ID            float64
Gender              object
Age                float64
EstimatedSalary    float64
Purchased          float64
dtype: object


In [8]:
# Check for missing values
print("Missing Values:")
print(df.isnull().sum())
print("\nTotal missing values:", df.isnull().sum().sum())

Missing Values:
User ID            0
Gender             0
Age                0
EstimatedSalary    0
Purchased          0
dtype: int64

Total missing values: 0


**Observation:** The dataset has 400 rows and 5 columns. There are no missing values. The features include `Gender` (categorical), `Age`, `EstimatedSalary` (numeric), and the target variable is `Purchased`.

---
### Step 2 — Data Preprocessing

In [9]:
from sklearn.preprocessing import LabelEncoder

# Step 2a: Handle missing values — none found, but showing the standard approach
df = df.dropna()  # No effect here, but good practice

# Step 2b: Encode categorical variable — Gender (Male=1, Female=0)
le = LabelEncoder()
df['Gender'] = le.fit_transform(df['Gender'])

print("Gender encoding — classes:", list(le.classes_))
print("Encoded values:  Female=0, Male=1")
df.head()

Gender encoding — classes: ['Female', 'Male']
Encoded values:  Female=0, Male=1


,User ID,Gender,Age,EstimatedSalary,Purchased
0,15624510.0,1,19.0,19000.0,0.0
1,15810944.0,1,35.0,20000.0,0.0
2,15668575.0,0,26.0,43000.0,0.0
3,15603246.0,0,27.0,57000.0,0.0
4,15804002.0,1,19.0,76000.0,0.0


In [10]:
# Step 2c: Select relevant features and separate X and y
X = df[['Age', 'EstimatedSalary']]   # Features
y = df['Purchased']                  # Target

print("Features (X):")
print(X.head())
print("\nTarget (y):")
print(y.head())
print("\nTarget value counts:")
print(y.value_counts())

Features (X):
    Age  EstimatedSalary
0  19.0          19000.0
1  35.0          20000.0
2  26.0          43000.0
3  27.0          57000.0
4  19.0          76000.0

Target (y):
0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: Purchased, dtype: float64

Target value counts:
Purchased
0.0    257
1.0    143
Name: count, dtype: int64


**Note:** As per the assignment, we use `Age` and `EstimatedSalary` as features. `User ID` is an identifier and is excluded. `Gender` was encoded but not included in the final feature set, keeping the model to 2D features for easier visualization.

---
### Step 3 — Train-Test Split (80/20)

In [11]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("Training set size: ", X_train.shape)
print("Testing set size:  ", X_test.shape)

Training set size:  (320, 2)
Testing set size:   (80, 2)


---
### Step 4 — Feature Scaling (Standard Scaler)

In [12]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

# Fit on training data ONLY, then transform both sets
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled  = scaler.transform(X_test)

print("Before scaling — Age mean: {:.2f}, std: {:.2f}".format(
    X_train['Age'].mean(), X_train['Age'].std()))

print("After scaling  — Age mean: {:.4f}, std: {:.4f}".format(
    X_train_scaled[:, 0].mean(), X_train_scaled[:, 0].std()))

Before scaling — Age mean: 37.87, std: 10.21
After scaling  — Age mean: -0.0000, std: 1.0000


**Why fit only on training data?** Fitting the scaler on the test set would cause *data leakage* — the model would indirectly learn statistics from test data. We always fit on train and apply to test.

---
### Step 5 — Model Training (Logistic Regression)

In [13]:
from sklearn.linear_model import LogisticRegression

# Initialise and train the model
model = LogisticRegression(random_state=42)
model.fit(X_train_scaled, y_train)

print("Model trained successfully!")
print("Coefficients:", model.coef_)
print("Intercept:   ", model.intercept_)

Model trained successfully!
Coefficients: [[1.96509784 1.13509021]]
Intercept:    [-0.99504916]


In [14]:
# Quick sanity check — training accuracy
train_accuracy = model.score(X_train_scaled, y_train)
print(f"Training Accuracy: {train_accuracy * 100:.2f}%")

Training Accuracy: 84.06%


---
## Summary — Part A Pipeline

| Step | Action | Output |
|------|--------|--------|
| 1 | Data Loading & Exploration | 400 rows, 5 cols, no nulls |
| 2 | Preprocessing | Encoded Gender, selected Age & EstimatedSalary |
| 3 | Train-Test Split (80/20) | 320 train / 80 test |
| 4 | Standard Scaling | Mean≈0, Std≈1 features |
| 5 | Logistic Regression Fit | Model ready for evaluation |